### DISCUSSION TITLE
update meta data using lua filter

### DISCUSSION THREAD

**Discussion by ThierryO** [quarto-dev/quarto-cli#9569] [Q&A] [answered]
### Description

Can we use a Lua filter to update the meta data? I'm trying to set the value of `book:body-footer`. The Lua filter below is my failing attempt.

```
return {
  {
    Meta = function(meta)
      meta.book.body-footer = 'test'
      return meta
    end
  }
}
```

**Comment by cscheid** [comment, marked as answer]
Did you follow the error message you're probably getting? `body-footer` is not valid Lua syntax in that context. You need `meta.book["body-footer"]`.

(In addition, this might not work depending on what you're trying to do with the metadata, because of the order in which documents are processed and filters are run matters.)

**Reply by mcanouil** [reply]
If it helps, here is a real life example:

- https://github.com/mcanouil/quarto-lua-env/blob/main/_extensions/lua-env/lua-env-filter.lua#L73-L88

**Comment by ThierryO** [comment]
Updating `meta.book["body-footer"]` did work. I can display the content with the shortcode `{{< meta book.body-footer>}}`. However it does not add the text as a footer to the pages. I worked around this by creating a shortcode and placing that in the `_quarto_yml`.

```
book:
  body-footer: '{{< footer >}}'
```

**Reply by mcanouil** [reply]
Why not use the variable shortcode? (https://quarto.org/docs/authoring/variables.html#var)

- `_variables.yml`

```yml
my-footer: Whatever
```

- `_quarto.yml`

```
book:
  body-footer: "{{< var my-footer >}}"

**Reply by ThierryO** [reply]
I'm in the process moving our [corporate identity](https://github.com/inbo/inbomd) from bookdown to quarto. The shortcode combines several meta items into the footer. Ideally the extension should always set the footer without the user having to set `book: body-footer`.

```
function is_empty(s)
  return s == nil or s == ''
end

return {
  ['footer'] = function(args, kwargs, meta)
    z = '<a rel="license" href="http://creativecommons.org/licenses/by/4.0/">' ..
      '<img alt="Creative Commons-License" style="border-width:0" class="ccby"/>' ..
      '</a>' .. meta.shortauthor
    if is_empty(meta.flandersqmd.year) then
      z = z .. '<h1 class = "missing">!!! Missing flandersqmd.year !!!</h1>'
    else
      z = z .. ' (' .. pandoc.utils.stringify(meta.flandersqmd.year) .. ')'
    end
    if is_empty(meta.flandersqmd.doi) then
      z = z .. '<h1 class = "missing">!!! Missing flandersqmd.doi !!!</h1>'
    else
      local x = pandoc.utils.stringify(meta.flandersqmd.doi)
      z = z .. ' <a href="https://doi.org/' .. x .. '">' .. x .. '</a>'
    end

    return pandoc.RawInline("html", z)
  end
}
```

**Reply by mcanouil** [reply]
Nice, thanks for the explanation!
